[TensorExpr] Support shape inference in TE for aten::cat.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42568 [TensorExpr] Disallow fallback to JIT interpreter from TensorExprKernel (flip the default).
* #42567 [TensorExpr] Apply GenericIntrinsicExpander recursively.
* #42566 [TensorExpr] Handle constant nodes in shape inference.
* **#42387 [TensorExpr] Support shape inference in TE for aten::cat.**

Differential Revision: [D22879281](https://our.internmc.facebook.com/intern/diff/D22879281)

Commit:
[TensorExpr] Support shape inference in TE for aten::cat.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Support shape inference in TE for aten::cat."


Differential Revision: [D22879281](https://our.internmc.facebook.com/intern/diff/D22879281)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Support shape inference in TE for aten::cat."


Differential Revision: [D22879281](https://our.internmc.facebook.com/intern/diff/D22879281)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Support shape inference in TE for aten::cat."


Differential Revision: [D22879281](https://our.internmc.facebook.com/intern/diff/D22879281)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -273,7 +273,29 @@ std::vector<ExprHandle> TensorExprKernel::inferSizesForValue(
+      // corresponding sizes of the input tensors, the other dimension have the
+      // same sizes.
+      auto const& n = v->node();
+      auto inputs = n->inputs()[0]->node()->inputs();
```
Does inputs()[0] necessarily have to have inputs?  What happens if it doesn't?

ZolotukhinM replied:
I thought that it would fire an assert, but after re-examination I found that it would just cause an out of bound access. I will change `[0]` to `.at(0)`, which has the boundary check. Thanks for noticing this!

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -273,7 +273,29 @@ std::vector<ExprHandle> TensorExprKernel::inferSizesForValue(
+      // same sizes.
+      auto const& n = v->node();
+      auto inputs = n->inputs()[0]->node()->inputs();
+      int64_t dim = n->inputs()[1]->node()->i(attr::value);
```
Does this assume the %dim is an integer constant?  What happens if it isn't?

ZolotukhinM replied:
Then we will fail miserably :) I will add some checks for this!

ZolotukhinM replied:
To expand a bit: when we add `aten::cat` nodes to a fusion group we do check that this is true: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L235-L243

I added an assert so that we fail loudly if something goes not like we expected.

bertmaher reviewed (approved):
I have a couple questions that I'd like to know the answers to, but otherwise this looks good.

Commit:
Update on "[TensorExpr] Support shape inference in TE for aten::cat."


Differential Revision: [D22879281](https://our.internmc.facebook.com/intern/diff/D22879281)

[ghstack-poisoned]
